### PR TITLE
Added API endpoint tests

### PR DESCRIPTION
**Pull request by iammosespaulr**
screenshot from 2018-01-16 11-55-41
When Server Isn't running
screenshot from 2018-01-16 11-56-09
When Server Is Running

### CODE REVIEW

**Comment by iammosespaulr** [comment]
@agentmilindu please review

**Comment by iammosespaulr** [comment]
@malithsen & @agentmilindu please review the pr and also my GCI task https://codein.withgoogle.com/dashboard/task-instances/5836999971831808/

**Comment by malithsen** [comment]
@iammosespaulr Good job in integrating unit tests. However the test cases could be better. Testing for a non existing end point isn't useful.
For starters, let's check `api/login` 
given a correct username and password (rand, rush) it should return auth level in body and a token in header.
Given incorrect login credentials it should return a 403 HTTP status

**Comment by iammosespaulr** [comment]
Sure will do the same ...

**Comment by iammosespaulr** [comment]
I have added login tests for auth:0 as well  the screenshots are posted in the next comment @agentmilindu @malithsen please review

**Comment by iammosespaulr** [comment]
For Correct Login Details
screenshot from 2018-01-16 19-12-06
For Incorrect Login Details
screenshot from 2018-01-16 19-12-23
@malithsen the changes have been made .. please review my GCI task as well

**Comment by iammosespaulr** [comment]
Added the 403 error check for incorrect login as well 
the screenshot

screenshot from 2018-01-16 19-27-24

**Comment by iammosespaulr** [comment]
@malithsen @agentmilindu please review  ... (GCI task : https://codein.withgoogle.com/dashboard/task-instances/5836999971831808/)

File: test/test_Bassa_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+'Accept-Encoding': 'gzip, deflate',
+'Referer': 'http://localhost:3000/',
+'Content-Type': 'application/x-www-form-urlencoded',
+'Content-Length': '28',
+'Origin': 'http://localhost:3000',
+'Connection': 'close'
+}
+payload = {''}
+class TestFlaskAPIUsingRequests(unittest.TestCase):
+    def test_api_for_404(self):
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Let's remove this test as it doesn't test for any functionality

File: test/test_Bassa_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        response = requests.get('http://localhost:5000')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_path_for_404(self):
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Remove this as well

File: test/test_Bassa_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        response = requests.get('http://localhost:5000/api')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_login(self):
+        resp = requests.post('http://localhost:5000/api/login',"user_name=rand&password=pass",headers=headers)
+        print(resp.status_code)
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Remove the unnecessary print statements

File: test/test_Bassa_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    def test_api_for_404(self):
+        response = requests.get('http://localhost:5000')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_path_for_404(self):
+        response = requests.get('http://localhost:5000/api')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_login(self):
+        resp = requests.post('http://localhost:5000/api/login',"user_name=rand&password=pass",headers=headers)
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Keep username and password in variables

File: test/test_Bassa_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        print(resp.status_code)
+	print(resp.json())
+        self.assertEqual(resp.json(),{u'auth': u'0'})
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: We need to test if the response have the token in the header. Let's check that in a separate test case

File: test/test_Bassa_endpoint.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+class TestFlaskAPIUsingRequests(unittest.TestCase):
+    def test_api_for_404(self):
+        response = requests.get('http://localhost:5000')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_path_for_404(self):
+        response = requests.get('http://localhost:5000/api')
+        print "code:",response.status_code
+        self.assertEqual(response.status_code,404)
+    def test_api_login(self):
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: When writing test cases it's good have descriptive function names which describes what the test case tests for.
eg: test_api_login_returns_auth_level(self)

https://stackoverflow.com/questions/1726622/how-should-unit-tests-be-documented

File: test/test_Bassa_endpoint.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import unittest
+import requests
+
+headers = {
+'Host': 'localhost:5000',
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Although all these headers are returned in browser request, all of them are not needed for the backend to function. Let's keep `Accept`, `Content-Type`, `Host`, `Origin` and remove the rest

**Comment by malithsen** [comment]
Also let's not commit `__pycache__` directory. Ideally it should have been ignored from our `.gitignore`

**Comment by iammosespaulr** [comment]
Ok @malithsen i will do the changes

**Comment by iammosespaulr** [comment]
@malithsen i have done the requested changes

File: test/test_Bassa_endpoint.py (hunk 7 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done ..

File: test/test_Bassa_endpoint.py (hunk 1 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py (hunk 2 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py (hunk 6 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py (hunk 4 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py (hunk 3 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py (hunk 5 above)
Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: done

File: test/test_Bassa_endpoint.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+correct_password="pass"
+incorrect_username="admin"
+incorrect_password="admin"
+correct_string="user_name="+correct_username+"&password="+correct_password
+incorrect_string="user_name="+incorrect_username+"&password="+incorrect_password
+payload = {''}
+class TestFlaskAPIUsingRequests(unittest.TestCase):
+    def test_api_login_returns_auth_level(self):
+        resp = requests.post('http://localhost:5000/api/login',correct_string,headers=headers)
+        def check_valid_auth(self):
```
Comment (malithsen) [inline test/test_Bassa_endpoint.py]: Why do we have a function inside a function?

Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: i wanted to seperate the login auth as another function just like you asked

Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: How else do I separate the check auth function as the response is received only after the POST...

Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: btw i have submitted my task for review as well ([Link](https://codein.withgoogle.com/dashboard/task-instances/5836999971831808/)) so you could review that as well

Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: also could you please ping me in gitter when you come online

Comment (malithsen) [reply test/test_Bassa_endpoint.py]: Ah Cool. Shall we have proper indentation for the following line

Comment (iammosespaulr) [reply test/test_Bassa_endpoint.py]: @malithsen i see that you are online .. could you come on gitter as  i have a few queries regarding this project

**Comment by malithsen** [comment]
@iammosespaulr Let's move the tests to components/core/tests directory

**Comment by iammosespaulr** [comment]
@malithsen Done ...  :smile:

**Comment by iammosespaulr** [comment]
Moved this PR to another pr  #412

**Pull request closed without merging**